### ISSUE TITLE
Train Yolov8 with txt images path

### ISSUE THREAD

**Issue by camiloromers** [ultralytics/ultralytics#3087]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I would like a clear example of how to pass a text file with relative paths for images and labels instead of directories when training YOLO models. According to the [documentation](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/?h=custom+data#train-on-custom-data), it is possible, but I am still unable to do it.

1.- First structure:

In dir images/**train.txt**
**train.txt** contains all relative images path img_1.jpg, img_2.jpg, img_3.jpg

1.- Second structure:

In dir images/train
Create multiple **txt** files where each file contains the relative img path.

Neither of them worked for me, and I would really appreciate any help.

**Comment by bit-scientist** [comment]
@camiloromers as for the second structure, you create  a `datasets` folder inside `ultralytics` project directory. And, let's say you have your images (.png or .jpg) and labels (.txt) inside your dataset folder called `datasetA`. You create `images` and `labels` folders for each split (train, valid and test) inside `datasetA` folder and put your images and labels files inside, respectively. The path in datasets folder would be:
```
|-- datasetA
|   |-- train
|       |-- images
|       |-- labels
|   |-- valid
|       |-- images
|       |-- labels
|   |-- test
|       |-- images
|       |-- labels
```
---

Configure the data.yaml file as follows:

```
train:
 - D:\ultralytics\datasets\datasetA\train\images

val:
 - D:\ultralytics\datasets\datasetA\valid\images

test:
 - D:\ultralytics\datasets\datasetA\test\images

nc: x # x here represents the number of classes.

names: [ 'person', 'bicycle', 'car', 'motorcycle', ...] 
```
The files in `labels` will be automatically found and dealt with. Good luck!

**Comment by camiloromers** [comment]
my question was more I can use txt file for images and labels instead of having images directory. I mean the txt file should point to the relative image path. Is that possible with yolo?

**Comment by glenn-jocher** [comment]
@camiloromers yes, it is possible to use a text file that contains relative paths for images and labels instead of having an image directory in YOLOv8. 

One way to accomplish this is to create a separate text file for each split (train, valid, and test) and include the relative paths for the images and labels in each file. In the `data.yaml` file, you would then reference the path to the folder containing the text files for each split instead of the image directory.

Another approach is to create a single text file that lists the relative paths for all images and labels and reference this file in the `data.yaml` file. When training, YOLOv8 will automatically read the labels for each image and create a corresponding `.pt` file containing the labels.

In either case, make sure the folder structure is correct and the relative paths in the text files point to the correct image and label files.

**Comment by camiloromers** [comment]
@glenn-jocher Is the following a valid structure?:

```
.
├── images
│   ├── train
│   │   └── train.txt
│   └── val
│       └── val.txt
├── labels
│   ├── train
│   │   └── train.txt
│   └── val
│       └── val.txt
```

**Comment by glenn-jocher** [comment]
Hello @camiloromers, thanks for reaching out to the YOLOv8 community! 

Regarding your folder structure, it appears to be a valid structure for storing training and validation images and labels. In this structure, you can create separate text files for each split (train and val), where each text file contains the relative paths for both the images and their corresponding labels.

For example, in the `train.txt` file, you would place the relative paths for the training images and their corresponding labels. The same would be done for the `val.txt` file with your validation images.

In the other approach, you can create a single text file which lists the relative paths for all images and their corresponding label files. This text file can then be referenced in your `data.yaml` file where the image and label data is defined for YOLOv8.

Regardless of which approach you take, make sure that the relative paths in your text files point to the correct image and label files and that the path to the folder containing the text files is correctly specified in the `data.yaml` file.

I hope this helps, if you have any further questions don't hesitate to ask!

**Comment by camiloromers** [comment]
@glenn-jocher Is it possible if you provide an example of second approach with just 3 images. My project has a two folders:

Images: img1.jpg, img2.jpg, img3.jpg
Labels: img1.txt, img2.txt, img3.txt

And I want to create a train.txt with first 2 images and val.txt with the last one.

Really appreciated.

**Comment by agvdndor** [comment]
> @glenn-jocher Is the following a valid structure?:
> 
> ```
> .
> ├── images
> │   ├── train
> │   │   └── train.txt
> │   └── val
> │       └── val.txt
> ├── labels
> │   ├── train
> │   │   └── train.txt
> │   └── val
> │       └── val.txt
> ```
> 
> If that is correct, the same txt file should have the reference for images and labels? how?
> 
> And may I get an example of the second approach please?

When looking at the code this does not seem to be possible.

It looks like the YoloDataset indeed has no problem loading the images from a txt file containing relative (or global) file paths as can bee seen in the [get_img_files ](https://github.com/ultralytics/ultralytics/blob/9d1e5567de48453f168013ff1032810bd95d39fe/ultralytics/yolo/data/base.py#L98) method.
```
def get_img_files(self, img_path):
        """Read image files."""
        try:
            f = []  # image files
            for p in img_path if isinstance(img_path, list) else [img_path]:
                p = Path(p)  # os-agnostic
                if p.is_dir():  # dir
                    f += glob.glob(str(p / '**' / '*.*'), recursive=True)
                    # F = list(p.rglob('*.*'))  # pathlib
                elif p.is_file():  # file
                    with open(p) as t:
                        t = t.read().strip().splitlines()
                        parent = str(p.parent) + os.sep
                        f += [x.replace('./', parent) if x.startswith('./') else x for x in t]  # local to global path
                        # F += [p.parent / x.lstrip(os.sep) for x in t]  # local to global path (pathlib)
                else:
                    raise FileNotFoundError(f'{self.prefix}{p} does not exist')
            im_files = sorted(x.replace('/', os.sep) for x in f if x.split('.')[-1].lower() in IMG_FORMATS)
            # self.img_files = sorted([x for x in f if x.suffix[1:].lower() in IMG_FORMATS])  # pathlib
            assert im_files, f'{self.prefix}No images found'
        except Exception as e:
            raise FileNotFoundError(f'{self.prefix}Error loading data from {img_path}\n{HELP_URL}') from e
        if self.fraction < 1:
            im_files = im_files[:round(len(im_files) * self.fraction)]
        return im_files
```

But the problem is that it will still [infer](https://github.com/ultralytics/ultralytics/blob/9d1e5567de48453f168013ff1032810bd95d39fe/ultralytics/yolo/data/utils.py#L39) the path to the label files from the image paths so any txt files under `labels/train/` or `labels/val` will be ignored:

```
def img2label_paths(img_paths):
    """Define label paths as a function of image paths."""
    sa, sb = f'{os.sep}images{os.sep}', f'{os.sep}labels{os.sep}'  # /images/, /labels/ substrings
    return [sb.join(x.rsplit(sa, 1)).rsplit('.', 1)[0] + '.txt' for x in img_paths]
```

This should be addressed so the relative or global paths in `labels/train/train.txt` and `labels/val/val.txt` are respected.

**Comment by JeromeC-B** [comment]
> @glenn-jocher Is it possible if you provide an example of second approach with just 3 images. My project has two folders:
> 
> ```
> ├── images
> │   ├── img_1.jpg
> │   ├── img_2.jpg
> │   ├── img_3.jpg
> ├── labels
> │   ├── img_1.txt
> │   ├── img_2.txt
> │   ├── img_3.txt
> ```
> 
> And I want to create a train.txt with first 2 images and val.txt with the last one.
> 
> Really appreciated.

Hi!:) Did you find a solution for this?:)

**Comment by camiloromers** [comment]
> > Is it possible if you provide an example of second approach with just 3 images. My project has two folders:
> > ```
> > ├── images
> > │   ├── img_1.jpg
> > │   ├── img_2.jpg
> > │   ├── img_3.jpg
> > ├── labels
> > │   ├── img_1.txt
> > │   ├── img_2.txt
> > │   ├── img_3.txt
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And I want to create a train.txt with first 2 images and val.txt with the last one.
> > Really appreciated.
> 
> @camiloromers Hi!:) Did you find a solution for this?:)

@JeromeC-B Nope :( I wanted an example of second approach.

**Comment by glenn-jocher** [comment]
@camiloromers Here's an example of how you can create the train.txt and val.txt files for the second approach using the folder structure you provided:

For the train.txt file, you would put the relative paths of the first two images (`img_1.jpg` and `img_2.jpg`) in the `images/train` folder, along with their corresponding label files (`img_1.txt` and `img_2.txt`) in the `labels/train` folder.

Contents of train.txt:
```
images/train/img_1.jpg
images/train/img_2.jpg
```

For the val.txt file, you would put the relative path of the third image (`img_3.jpg`) in the `images/val` folder, along with its corresponding label file (`img_3.txt`) in the `labels/val` folder.

Contents of val.txt:
```
images/val/img_3.jpg
```

Make sure the txt files are saved in the `images/train` and `images/val` folders, respectively.

By referencing these train.txt and val.txt files in your `data.yaml` file for YOLOv8, the YOLOv8 training process will use these files to locate the images and their corresponding labels during training and validation.

I hope this helps! Let me know if you have any further questions.

**Comment by camiloromers** [comment]
@glenn-jocher you don't have to specify annotated txt files for every txt file you generate? Because I can see only jpg images are referenced on train.txt and val.txt

**Comment by rromanchuk** [comment]
**My requirements**
- There is no concept of a global datasets directory
- The name of the dataset and it's location on the filesystem is dynamically generated, absolute paths can't be hard coded anywhere
- Splits logic should not depend on directory organization
- Modifying splits should not require an image to change location
- Modifying split data should not require an entire re-upload of image data
- Modifying splits should not change the last modified of the backing resource
- An image included in the dataset may not be included in any train/val/test splits
- The trainer consuming the dataset should have read-only access. Cache, artifacts, plots, models, runs should not be dropped anywhere inside the dataset directory
- The dataset directory should be considered ephemeral and frozen
-  The location of script execution, the CWD, on the file system, is not related to the location of the dataset, and cannot be accessed with simple `../` traversal
- `data.yaml` is considered frozen, and version controlled, it cannot be manipulated at runtime
- `settings.yaml` cannot be modified at runtime

The only way i have gotten some of these requirements is by hacking around ultralytics injection. Detection only, classification looks to be SOL 

**Hacks required**
- the string `images` and `labels` are not arbitrary. Labels are not configurable but depend on dir structure of a path named "images", even when providing .txt list
- CWD must be root of the dataset in order to use relative-only references that can be used to hide from settings.yaml hijacking
- `data=` must include absolute path to avoid setting.yaml hijacking
- because execution is now taking placing in CWD of dataset, via a `chdir`,  the relative `weights`, `runs` are now relative to CWD different from original script execution
- The arbitrary absolute path of the CWD of script execution should be saved before chdir, and once again, you can prevent setting.yaml from hijacking, by using the overloaded and poorly named `project=` and `name=` attributes to assign the arbitrary, runtime dynamic, absolute path to your original CWD so artifacts aren't polluting a frozen and ephemeral datasets directory

Example

`ls /this/absolute/path/is/resolved/at/runtime/for/this/dynamically-generated-name-and-location-of-dataset-to-train-with`
```
├ data.yaml
├ train.txt
├ val.txt
├ test.txt
├── images
│   ├── img_1.jpg
│   ├── img_2.jpg
│   ├── img_3.jpg
│   ├── img_4.jpg
├── labels
│   ├── img_1.txt
│   ├── img_2.txt
│   ├── img_3.txt
│   ├── img_4.txt
```
`pwd`
/this/cwd/is/dynamically/assigned/at/runtime

train.txt
```txt
images/img_1.jpg
images/img_4.jpg
```

val.txt
```txt
images/img_2.jpg
```

test.txt
```txt
images/img_3.jpg
```
data.yaml
```yaml
train: train.txt
val: val.txt
test: test.txt

nc: 4
names: ['ultralytics', 'makes', 'obnoxious', 'assumptions']
```

settings.yaml
```yaml
datasets_dir: datasets
weights_dir: weights
runs_dir: runs
```

Running
```python
from clearml import Task, Dataset
from ultralytics import YOLO
import argparse
import os
parser = argparse.ArgumentParser(description='')
parser.add_argument('--epochs', type=int, default=300,
                    help='Number of epochs to train (default = %(default)s)')
parser.add_argument('--batch', type=int, default=64,
                    help='Batch size(default = %(default)s)')
parser.add_argument('--imgsz', type=int, default=640,
                    help='Image size (default = %(default)s)')

dataset_path = Dataset.get(
    dataset_id=user_args.dataset,
    alias="my alias"
).get_local_copy()

# WORKING DIRECTORY MUST BE dataset_path
os.chdir(dataset_path)
# WORKING DIRECTORY MUST BE dataset_path

# Artifacts now incorrectly saved to `dataset_path`
model = YOLO(user_args.model)
model.train(data=f"{dataset_path}/data.yaml", epochs=user_args.epochs, imgsz=user_args.imgsz, batch=user_args.batch, optimizer=user_args.optimizer, save_period=10, save=True)
```

**Comment by glenn-jocher** [comment]
@rromanchuk thank you for posting your detailed requirements. Since YOLOv8 is still under development, we appreciate any feedback, use cases, and suggestions users like you can provide. We will work towards making the dataset handling more flexible and adjustable to various data architectures.

Our current implementation indeed makes some assumptions about your directory structure with images and labels being in the same directory. We will further explore ways to provide customization while keeping simplicity for users who are new to this field. 

"train.txt", "val.txt" and "test.txt" files are identifiers for which images are utilized in each stage of the training process. These should include the paths to images, but the dataset itself does not directly dictate their location. We understand your requirement about keeping the dataset frozen, and we will account for this in future updates.

Regarding the dynamic dataset paths, we acknowledge that there is room for improvement. We aim to make our software more flexible to accommodate various directory structures and locations. We have noted all your requirements and hopefully they will be reflected in future updates.

Thank you again for your detailed feedback, it's extremely helpful to us. We aim to make YOLOv8 as user-friendly and adaptable as possible, and your input is very valuable in this regard. Stay tuned for upcoming updates.

**Comment by s0966066980** [comment]
"Can the following format be used in a txt file?
image_path x_min, y_min, x_max, y_max, class_id x_min, y_min, ..., class_id"

**Comment by glenn-jocher** [comment]
@s0966066980 no, the format you've described is not supported. The correct format for a label `.txt` file in YOLOv8 is one line per object, each with five space-separated values: class_id, x_center, y_center, width, and height, all normalized to [0, 1] by the image width and height.
